wasm/src/main/scala/ir2wasm/WasmExpressionBuilder.scala

```diff
@@ -573,46 +821,150 @@ private class WasmExpressionBuilder private (
-        case IRTypes.StringType =>
-          () // no-op
-
-        case IRTypes.BooleanType =>
-          instrs += IF(BlockType.ValueType(wasmStringType))
-          genLiteral(IRTrees.StringLiteral("true")(tree.pos))
-          instrs += ELSE
-          genLiteral(IRTrees.StringLiteral("false")(tree.pos))
-          instrs += END
+      def genWithDispatch(isAncestorOfHijackedClass: Boolean): Unit = {
+        /* Somewhat duplicated from genApplyNonPrim, but specialized for
+         * `toString`, and where the handling of `null` is different.
+         *
+         * We need to return the `"null"` string in two special cases:
+         * - if the value itself is `null`, or
+         * - if the value's `toString(): String` method returns `null`!
+         */
+
+        // A local for a copy of the receiver that we will use to resolve dispatch
+        val receiverLocalForDispatch: WasmLocalName = {
+          val name = fctx.genSyntheticLocalName()
+          val typ = Types.WasmRefType(Types.WasmHeapType.ObjectType)
+          fctx.locals.define(WasmLocal(name, typ, isParameter = false))
+          name
+        }
 
-        case IRTypes.CharType =>
-          instrs += WasmInstr.ARRAY_NEW_FIXED(
-            TypeIdx(WasmTypeName.WasmArrayTypeName.stringData),
-            I32(1)
+        val objectClassInfo = ctx.getClassInfo(IRNames.ObjectClass)
+
+        if (!isAncestorOfHijackedClass) {
+          /* Standard dispatch codegen, with dedicated null handling.
+           *
+           * The overall structure of the generated code is as follows:
+           *
+           * block (ref any) $done
+           *   block $isNull
+           *     load receiver as (ref null java.lang.Object)
+           *     br_on_null $isNull
+           *     generate standard table-based dispatch
+           *     br_on_non_null $done
+           *   end $isNull
+           *   gen "null"
+           * end $done
+           */
+
+          val labelDone = fctx.genLabel()
+          instrs += BLOCK(BlockType.ValueType(Types.WasmRefType.any), Some(labelDone))
+
+          val labelIsNull = fctx.genLabel()
+          instrs += BLOCK(BlockType.ValueType(), Some(labelIsNull))
+
+          genTreeAuto(tree)
+          instrs += BR_ON_NULL(labelIsNull)
+          instrs += LOCAL_TEE(LocalIdx(receiverLocalForDispatch))
+          genTableDispatch(objectClassInfo, toStringMethodName, receiverLocalForDispatch)
+          instrs += BR_ON_NON_NULL(labelDone)
+
+          instrs += END // block $isNull
+
+          genLiteral(IRTrees.StringLiteral("null")(tree.pos))
+
+          instrs += END // block $done
+        } else {
+          /* Dispatch where the receiver can be a JS value.
+           *
+           * The overall structure of the generated code is as follows:
+           *
+           * block (ref any) $done
+           *   block anyref $notOurObject
+           *     load receiver
+           *     br_on_cast_fail anyref (ref $targetRealClass) $notOurObject
```

Comment:
It will succeed for `long` and `char`, as well as for any real class. We are in a situation where the receiver's static type is an *ancestor* of a hijacked class (e.g., `java.lang.CharSequence`, or even `any`). So at run-time we can have either primitives (possibly boxed for `long` and `char`) or real class instances.